src/crimson/osd/replicated_recovery_backend.cc

```diff
@@ -694,6 +694,16 @@ ReplicatedRecoveryBackend::_handle_pull_response(
         recovery_waiter.obc = obc;
         obc->obs.oi.decode_no_oid(push_op.attrset.at(OI_ATTR), push_op.soid);
         pull_info.recovery_info.oi = obc->obs.oi;
+
+        auto ss_attr_iter = push_op.attrset.find(SS_ATTR);
+        if (ss_attr_iter != push_op.attrset.end()) {
+          if (!obc->ssc) {
+            obc->ssc = new crimson::osd::SnapSetContext(
+              push_op.soid.get_snapdir());
+          }
+          obc->ssc->snapset = SnapSet(ss_attr_iter->second);
+          obc->ssc->exists = true;
+        }
```

Comment:
We load the obc using pg's `obc_loader`. The issue here is that obc_loader doesn't support additional attrset -  it will always use PGBackend::store attrs. Classic handles that by allowing to pass custom attrs to `get_snapset_context`.
I would suggest editing `ObjectContextLoader::with_obc` to support using custom attrs (push_op's in this case).
Later on, while handling pull response we could assert that the obc in question shares the same SnapSet with the one found in the push_op's attrset:
```
assert(!obc->ssc->exists || ss.seq  == obc->ssc->snapset.seq);
```

Reply:
Yeah, will do.

Reply:
> I'm afraid that creating the ssc here breaks obc_loader logic.

Hi, matan, it looks that the only use case of passing custom attrs to `get_object_context`/`get_snapset_context` is in the pull response handling, and the current obc_loader is implemented tolerating getting an empty obc: 
https://github.com/ceph/ceph/blob/0f539cd0113d7d09a77172608f45c48903ec9d9a/src/crimson/osd/pg_backend.cc#L137C21-L148

So I think maybe it's better we set attrs here, what do you think?

Reply:
I still prefer changing the obc_loader, although this is shouldn't be a blocker. The SnapSet ctor would need to be part of a try block if we end up with the current implementation. https://github.com/ceph/ceph/pull/53281#discussion_r1315678488